openedx/features/content_tagging/rest_api/v1/views.py

```diff
@@ -0,0 +1,163 @@
+        * description (optional): Provides extra information for the user when applying tags from this taxonomy to an object.
+        * enabled (optional): Only enabled taxonomies will be shown to authors.
+        * required (optional): Indicates that one or more tags from this taxonomy must be added to an object.
+        * allow_multiple (optional): Indicates that multiple tags from this taxonomy may be added to an object.
+        * allow_free_text (optional): Indicates that tags in this taxonomy need not be predefined; authors may enter their own tag values.
+
+    **Update Example Requests**
+        PUT api/tagging/v1/taxonomy/:id                                            - Update a taxonomy
+        {
+            "name": "Taxonomy New Name",
+            "description": "This is a new description",
+            "enabled": False,
+            "required": False,
+            "allow_multiple": False,
+            "allow_free_text": True,
+        }
+        PATCH api/tagging/v1/taxonomy/:id                                          - Partially update a taxonomy
+        {
+            "name": "Taxonomy New Name",
+        }
+
+    **Update Query Returns**
+        * 200 - Success
+        * 403 - Permission denied
+
+    **Delete Parameters**
+        * id (required): - The id of the taxonomy to delete
+
+    **Delete Example Requests**
+        DELETE api/tagging/v1/taxonomy/:id                                         - Delete a taxonomy
+
+    **Delete Query Returns**
+        * 200 - Success
+        * 404 - Taxonomy not found
+        * 403 - Permission denied
+
+    """
+
+    def get_queryset(self):
+        """
+        Return a list of taxonomies.
+
+        Returns all taxonomies by default.
+        If you want the disabled taxonomies, pass enabled=False.
+        If you want the enabled taxonomies, pass enabled=True.
+        """
+        query_params = TaxonomyOrgListQueryParamsSerializer(
+            data=self.request.query_params.dict()
+        )
+        query_params.is_valid(raise_exception=True)
+        enabled = query_params.validated_data.get("enabled", None)
+        org = query_params.validated_data.get("org", None)
+        taxonomies = get_taxonomies_for_org(enabled, org, False)
+
+        if is_global_taxonomy_admin(self.request.user):
+            return taxonomies
+        else:
+            # This should be an optional parameter on get_taxonomies_for_org?
+            user_orgs = get_user_taxonomy_orgs(self.request.user)
+            user_orgs_short_name = [org.short_name for org in user_orgs]
+            return taxonomies.filter(
+            Q(
+                Exists(
+                    TaxonomyOrg.objects.filter(
+                        taxonomy=OuterRef("pk"),
+                        rel_type=TaxonomyOrg.RelType.OWNER,
+                        org__short_name__in=user_orgs_short_name,
+                    )
+                )
+            )
+            | Q(enabled=True)
+        )
```

Comment:
This code probably needs to go to the API.
`get_taxonomies_for_org` should have a parameter (`from_orgs`?) that receives a list of orgs to which the caller has permissions.

We need this to differentiate both user cases
```python
("userA", None, "orgA", ("st1", "t1", "tA1", "tA2", "tC1", "tC2")) # The user can see "tA2" because he is from A
("userB", None, "orgA", ("st1", "t1", "tA1", "tC1", "tC2")) # "tA2" is filtered out because is disabled
```

Reply:
> This code probably needs to go to the API.

Agreed.. shouldn't have complex model queries in the views here.

Is there a way to use the `rules` to filter these instead of reproducing all of that logic here? Something like:

```
if "org" in params:
    taxonomies = get_taxonomies_for_org(enabled, org)
else:
    taxonomies = get_taxonomies(enabled, org)

returned = []
for taxonomy in taxonomies[paginated]:
    if request.user.has_perm("oel_tagging.view_taxonomy", taxonomy):
        returned.append(taxonomy)
```

Reply:
I don't think we will be able to. We want to filter before pagination (or we will end with "holes" in pages). Also, we don't want to iterate over the whole list before pagination (to know the record count, for example).

I will try to consolidate this and put it in `rules.py` as a sort of "taxonomy_list_filter". That way, we at least leave can_view_taxonomy and the list filter "together.

Reply:
The actual solutions is better proposed here: https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0009-bp-permissions.html